src/util/dateHelpers.js

```diff
@@ -39,26 +39,52 @@ export const numericToCalendar = (numDate) => {
 };
 
 /**
- * Convert a calendar date to a numeric one.
- * This function is meant to behave similarly to TreeTime's `numeric_date`
- * as found in v0.7*. Note that for negative dates, i.e. BCE, no fraction
- * in the year will be returned.
+ * Convert a YYYY-MM-DD string to a numeric date. This function is meant to
+ * behave similarly to TreeTime's `numeric_date` as found in v0.7*. For negative
+ * dates (i.e. BCE) we simply return the year (ignoring month / day). Ambiguity
+ * is optionally allowed in the form of YYYY-MM-XX or YYYY-XX-XX in which case
+ * the midpoint of the implied range is returned. All non compliant inputs
+ * return `undefined`.
  * @param {string} calDate in format YYYY-MM-DD
- * @returns {float} YYYY.F, where F is the fraction of the year passed
+ * @param {boolean} ambiguity
+ * @returns {float|undefined} YYYY.F, where F is the fraction of the year passed
  */
-export const calendarToNumeric = (calDate) => {
+export const calendarToNumeric = (calDate, ambiguity=false) => {
+  if (typeof calDate !== "string") return undefined;
   if (calDate[0]==='-') {
-    const pieces = calDate.substring(1).split('-');
-    return -parseFloat(pieces[0]);
+    const d = -parseFloat(calDate.substring(1).split('-')[0]);
+    return isNaN(d) ? undefined : d;
   }
-  /* Beware: for `Date`, months are 0-indexed, days are 1-indexed */
-  const [year, month, day] = calDate.split("-").map((n) => parseInt(n, 10));
+  const fields = calDate.split("-");
+  if (fields.length !== 3) return undefined;
+  const [year, month, day] = fields;
+  const [numYear, numMonth, numDay] = fields.map((d) => parseInt(d, 10));
+  
+  if (calDate.includes("X")) {
+    if (!ambiguity) return undefined
+    if (year.includes("X")) return undefined;
+    if (month.includes("X")) {
```

Comment:
As written, this allows `1970-X-XX` or `1970-XXX-XX` to work, while `1970-01-X` or `1970-01-XXX` would fail; check for exactly `XX` to match the spec.

```suggestion
    if (month==="XX")) {
```

Reply:
Fixed up + tests added